[RFC] Catchup with upstream quickfix patches

jamessan opened the pull request:
- [x] vim-patch:7.4.1752

Problem:    When adding to the quickfix list the current position is reset.
Solution:   Do not reset the position when not needed. (Yegappan Lakshmanan)

https://github.com/vim/vim/commit/c1808d5822ed9534ef7f0fe509b15bee92a5cc28


- [x] vim-patch:7.4.1768

Problem:    Arguments of setqflist() are not checked properly.
Solution:   Add better checks, add a test. (Nikolai Pavlov, Hirohito Higashi,
            closes vim/vim#661)

https://github.com/vim/vim/commit/d106e5ba7f10f0d2a14eaefe5d78405044416cb9


- [x] vim-patch:7.4.1802

Problem:    Quickfix doesn't handle long lines well, they are split.
Solution:   Drop characters after a limit. (Anton Lindqvist)

https://github.com/vim/vim/commit/6be8c8e165204b8aa4eeb8a52be87a58d8b41b9e


- [x] vim-patch:7.4.1813

Problem:    Memory access error when running test_quickfix.
Solution:   Allocate one more byte. (Yegappan Lakshmanan)

https://github.com/vim/vim/commit/9b4ebc692d77ca8ef90d72517347f74c2474dd3d


- [x] vim-patch:7.4.1815

Problem:    Compiler warnings for unused variables. (Ajit Thakkar)
Solution:   Add a dummy initialization. (Yasuhiro Matsumoto)

https://github.com/vim/vim/commit/9a3b3311d26c990208150255ad65472bb4eefaa4


- [x] vim-patch:7.4.1823

Problem:    Warning from 64 bit compiler.
Solution:   Add type cast. (Mike Williams)

https://github.com/vim/vim/commit/d9db8b448c214eb583e84c598bca0688b9202ba7


- [x] vim-patch:7.4.1841

Problem:    The code to reallocate the buffer used for quickfix is repeated.
Solution:   Move the code to a function. (Yegappan Lakshmanan, closes vim/vim#831)

https://github.com/vim/vim/commit/2b2b8ae5ab37b04584633c469265d85825166905


- [x] vim-patch:7.4.1871

Problem:    Appending to the quickfix list while the quickfix window is open
            is very slow.
Solution:   Do not delete all the lines, only append the new ones.  Avoid
            using a window while updating the list. (closes vim/vim#841)

https://github.com/vim/vim/commit/864293abb72d62604d8d6b458addfb43c14230c3


- [x] vim-patch:7.4.1881

Problem:    Appending to a long quickfix list is slow.
Solution:   Add qf_last.

https://github.com/vim/vim/commit/83e6d7ac6a1c2a0cb5ee6c8420a5dc792f1d5ffa


- [x] vim-patch:7.4.1882

Problem:    Check for line break at end of line wrong. (Dominique Pelle)
Solution:   Correct the logic.

https://github.com/vim/vim/commit/b37662a0fbb952838fca87aff4d26b596030b67b


- [x] vim-patch:7.4.1884

Problem:    Updating marks in a quickfix list is very slow when the list is
            long.
Solution:   Only update marks if the buffer has a quickfix entry.

https://github.com/vim/vim/commit/2f095a4bc4d786e0ac834f48dd18a94fe2d140e3


- [x] vim-patch:7.4.1912

Problem:    No test for using setqflist() on an older quickfix list.
Solution:   Add a couple of tests.

https://github.com/vim/vim/commit/1cee693b310e1494115a1677fac064941092e1bb


- [x] vim-patch:7.4.1937

Problem:    No test for directory stack in quickfix.
Solution:   Add a test. (Yegappan Lakshmanan)

https://github.com/vim/vim/commit/ab47c61f46f0797308217cd3c045472cb0ec3195


- [x] vim-patch:7.4.1941

Problem:    Not all quickfix tests are also done  with the location lists.
Solution:   Test more quickfix code.  Use user commands instead of "exe".
            (Yegappan Lakshmanan)

https://github.com/vim/vim/commit/3ef5bf7d459d6b8a21aaefc80f65448f5a7aa59f


- [x] vim-patch:7.4.1949

Problem:    Minor problems with the quickfix code.
Solution:   Fix the problems. (Yegappan Lakshmanan)

https://github.com/vim/vim/commit/38df43bd13a2498cc96b3ddd9a20dd75126bd171


- [x] vim-patch:7.4.1950

Problem:    Quickfix long lines test not executed for buffer.
Solution:   Call the function to test long lines. (Yegappan Lakshmanan)

https://github.com/vim/vim/commit/f50df3925b7c909d1cda4c868d8c7ba38189aabe


- [x] vim-patch:7.4.1953

Problem:    Not all parts of the quickfix code are tested.
Solution:   Add more tests. (Yegappan Lakshmanan)

https://github.com/vim/vim/commit/049cba9e9760152b5695399a991dc61cea9ba143


- [x] vim-patch:7.4.1964

Problem:    The quickfix init function is too big.
Solution:   Factor out parsing 'errorformat' to a separate function. (Yegappan
            Lakshmanan)

https://github.com/vim/vim/commit/688e3d1fd9b9129a5ba0e0d599ccfe6f4443daf3


- [x] vim-patch:7.4.1966

Problem:    Coverity reports a resource leak.
Solution:   Close "fd" also when bailing out.

https://github.com/vim/vim/commit/bcf7772a23624edc0942120e564f6b4ac95604ad


- [x] vim-patch:7.4.1980

Problem:    'errorformat' is parsed for every call to ":caddexpr".  Can't add
            to two location lists asynchronously.
Solution:   Keep the previously parsed data when appropriate. (mostly by
            Yegappan Lakshmanan)

https://github.com/vim/vim/commit/361c8f0e517e41f1f1d34dae328044406fde80ac


- [x] vim-patch:7.4.1984

Problem:    Not all quickfix features are tested.
Solution:   Add a few more tests. (Yegappan Lakshmanan)

https://github.com/vim/vim/commit/0fcc7c6dd1902b71e0e7d0a35ddabafef6455a83


- [x] vim-patch:7.4.1997

Problem:    Cannot easily scroll the quickfix window.
Solution:   Add ":cbottom".

https://github.com/vim/vim/commit/dcb170018642ec144cd87d9d9fe076575b8d1263


- [x] vim-patch:7.4.2010

Problem:    There is a :cbottom command but no :lbottom command.
Solution:   Add :lbottom. (Yegappan Lakshmanan)

https://github.com/vim/vim/commit/537ef08408c50e0c4104d57f74993b3b0ed9560d


- [x] vim-patch:8.0.0006

Problem:    ":lb" is interpreted as ":lbottom" while the documentation says it
            means ":lbuffer".
Solution:   Adjust the order of the commands. (haya14busa, closes vim/vim#1093)

https://github.com/vim/vim/commit/ebdd90ac282909c0bfcd6e83e70505abbbd5a38d

jamessan committed:
vim-patch:7.4.1752

Problem:    When adding to the quickfix list the current position is reset.
Solution:   Do not reset the position when not needed. (Yegappan Lakshmanan)

https://github.com/vim/vim/commit/c1808d5822ed9534ef7f0fe509b15bee92a5cc28

jamessan committed:
vim-patch:7.4.1768

Problem:    Arguments of setqflist() are not checked properly.
Solution:   Add better checks, add a test. (Nikolai Pavlov, Hirohito Higashi,
            closes vim/vim#661)

https://github.com/vim/vim/commit/d106e5ba7f10f0d2a14eaefe5d78405044416cb9

jamessan committed:
vim-patch:7.4.1802

Problem:    Quickfix doesn't handle long lines well, they are split.
Solution:   Drop characters after a limit. (Anton Lindqvist)

https://github.com/vim/vim/commit/6be8c8e165204b8aa4eeb8a52be87a58d8b41b9e

jamessan committed:
vim-patch:7.4.1813

Problem:    Memory access error when running test_quickfix.
Solution:   Allocate one more byte. (Yegappan Lakshmanan)

https://github.com/vim/vim/commit/9b4ebc692d77ca8ef90d72517347f74c2474dd3d

jamessan committed:
vim-patch:7.4.1815

Problem:    Compiler warnings for unused variables. (Ajit Thakkar)
Solution:   Add a dummy initialization. (Yasuhiro Matsumoto)

https://github.com/vim/vim/commit/9a3b3311d26c990208150255ad65472bb4eefaa4

jamessan committed:
vim-patch:7.4.1841

Problem:    The code to reallocate the buffer used for quickfix is repeated.
Solution:   Move the code to a function. (Yegappan Lakshmanan, closes vim/vim#831)

https://github.com/vim/vim/commit/2b2b8ae5ab37b04584633c469265d85825166905

jamessan committed:
vim-patch:7.4.1871

Problem:    Appending to the quickfix list while the quickfix window is open
            is very slow.
Solution:   Do not delete all the lines, only append the new ones.  Avoid
            using a window while updating the list. (closes vim/vim#841)

https://github.com/vim/vim/commit/864293abb72d62604d8d6b458addfb43c14230c3

jamessan committed:
vim-patch:7.4.1881

Problem:    Appending to a long quickfix list is slow.
Solution:   Add qf_last.

https://github.com/vim/vim/commit/83e6d7ac6a1c2a0cb5ee6c8420a5dc792f1d5ffa

jamessan committed:
vim-patch:7.4.1882

Problem:    Check for line break at end of line wrong. (Dominique Pelle)
Solution:   Correct the logic.

https://github.com/vim/vim/commit/b37662a0fbb952838fca87aff4d26b596030b67b

jamessan committed:
vim-patch:7.4.1884

Problem:    Updating marks in a quickfix list is very slow when the list is
            long.
Solution:   Only update marks if the buffer has a quickfix entry.

https://github.com/vim/vim/commit/2f095a4bc4d786e0ac834f48dd18a94fe2d140e3

jamessan committed:
vim-patch:7.4.1912

Problem:    No test for using setqflist() on an older quickfix list.
Solution:   Add a couple of tests.

https://github.com/vim/vim/commit/1cee693b310e1494115a1677fac064941092e1bb

jamessan committed:
vim-patch:7.4.1937

Problem:    No test for directory stack in quickfix.
Solution:   Add a test. (Yegappan Lakshmanan)

https://github.com/vim/vim/commit/ab47c61f46f0797308217cd3c045472cb0ec3195

jamessan committed:
vim-patch:7.4.1941

Problem:    Not all quickfix tests are also done  with the location lists.
Solution:   Test more quickfix code.  Use user commands instead of "exe".
            (Yegappan Lakshmanan)

https://github.com/vim/vim/commit/3ef5bf7d459d6b8a21aaefc80f65448f5a7aa59f

jamessan committed:
vim-patch:7.4.1949

Problem:    Minor problems with the quickfix code.
Solution:   Fix the problems. (Yegappan Lakshmanan)

https://github.com/vim/vim/commit/38df43bd13a2498cc96b3ddd9a20dd75126bd171

jamessan committed:
vim-patch:7.4.1950

Problem:    Quickfix long lines test not executed for buffer.
Solution:   Call the function to test long lines. (Yegappan Lakshmanan)

https://github.com/vim/vim/commit/f50df3925b7c909d1cda4c868d8c7ba38189aabe

jamessan committed:
vim-patch:7.4.1953

Problem:    Not all parts of the quickfix code are tested.
Solution:   Add more tests. (Yegappan Lakshmanan)

https://github.com/vim/vim/commit/049cba9e9760152b5695399a991dc61cea9ba143

jamessan committed:
vim-patch:7.4.1964

Problem:    The quickfix init function is too big.
Solution:   Factor out parsing 'errorformat' to a separate function. (Yegappan
            Lakshmanan)

https://github.com/vim/vim/commit/688e3d1fd9b9129a5ba0e0d599ccfe6f4443daf3

jamessan committed:
vim-patch:7.4.1966

Problem:    Coverity reports a resource leak.
Solution:   Close "fd" also when bailing out.

https://github.com/vim/vim/commit/bcf7772a23624edc0942120e564f6b4ac95604ad

jamessan committed:
vim-patch:7.4.1984

Problem:    Not all quickfix features are tested.
Solution:   Add a few more tests. (Yegappan Lakshmanan)

https://github.com/vim/vim/commit/0fcc7c6dd1902b71e0e7d0a35ddabafef6455a83

Shougo commented:
Great work!

Missing quickfix patches: 7.4.1823, 2017, 2018, 2046, 2061, 2064, 2067,2081, 2144,2151,2264, 2299

jamessan committed:
vim-patch:7.4.1823

Problem:    Warning from 64 bit compiler.
Solution:   Add type cast. (Mike Williams)

https://github.com/vim/vim/commit/d9db8b448c214eb583e84c598bca0688b9202ba7

Already silenced in earlier commits by using size_t.

jamessan committed:
test: quickfix: Move upstream vim quickfix test to src/nvim/testdir

Massaging the upstream patches for this test into the lua tests are too
cumbersome and slow down patching.

jamessan commented:
> Missing quickfix patches: 7.4.1823, 2017, 2018, 2046, 2061, 2064, 2067,2081, 2144,2151,2264, 2299

1823, 2064, 2067, 2081, 2144, 2151, 2264, and 2299 look good.

2017 and 2018 are more targeted at improving buffer lookup.

2046 and 2061 might be ok, as long as they don't overlap with the buffer lookup changes.

jamessan commented:
I'll handle patches up through 7.4.2010 with this PR, then get back to some other outstanding PRs.

jamessan committed:
vim-patch:7.4.1980

Problem:    'errorformat' is parsed for every call to ":caddexpr".  Can't add
            to two location lists asynchronously.
Solution:   Keep the previously parsed data when appropriate. (mostly by
            Yegappan Lakshmanan)

https://github.com/vim/vim/commit/361c8f0e517e41f1f1d34dae328044406fde80ac

jamessan committed:
vim-patch:7.4.1997

Problem:    Cannot easily scroll the quickfix window.
Solution:   Add ":cbottom".

https://github.com/vim/vim/commit/dcb170018642ec144cd87d9d9fe076575b8d1263

jamessan committed:
vim-patch:7.4.2010

Problem:    There is a :cbottom command but no :lbottom command.
Solution:   Add :lbottom. (Yegappan Lakshmanan)

https://github.com/vim/vim/commit/537ef08408c50e0c4104d57f74993b3b0ed9560d

jamessan committed:
vim-patch:8.0.0006

Problem:    ":lb" is interpreted as ":lbottom" while the documentation says it
            means ":lbuffer".
Solution:   Adjust the order of the commands. (haya14busa, closes vim/vim#1093)

https://github.com/vim/vim/commit/ebdd90ac282909c0bfcd6e83e70505abbbd5a38d

jamessan committed:
lint

jamessan commented:
Ok, I think this should be good.

Shougo commented:
Rest of the quickfix patches after Vim 8.0:
8.0017, 8.0026, 8.0037, 8.0078, 8.0079, 8.0084

jamessan commented:
Thanks, @Shougo.  I'll handle them in a separate PR, but having the list here will help when I get to that.

Shougo commented:
The tests are passed.
Why it is not merged?

justinmk commented on src/nvim/quickfix.c:
```diff
@@ -168,17 +176,272 @@ qf_init (
+  for (int round = FMT_PATTERNS - 1; round >= 0; ) {
+    i += STRLEN(fmt_pat[round--].pattern);
+  }
+#ifdef COLON_IN_FILENAME
```
FWIW, we never define this

jamessan replied:
Apparently Vim doesn't either.

justinmk commented on src/nvim/quickfix.c:
```diff
@@ -596,12 +791,23 @@ qf_init_ext (
+            errmsglen = linelen + 1;
+          }
+          STRLCPY(errmsg, linebuf, linelen + 1);
+        } else if ((i = (int)fmt_ptr->addr[5]) > 0) {             // %m
```
comment should be aligned with the other cases

justinmk commented:
If the tests fail correctly, should be safe to merge this.
